### ISSUE TITLE
[JIT] Incorrect gradient results under script mode

### ISSUE THREAD

**Issue by jermainewang** [pytorch/pytorch#32385]
## 🐛 Bug

I was playing with the torchscript example from the [tutorial](https://pytorch.org/tutorials/beginner/Intro_to_TorchScript_tutorial.html) and found that gradients may not be correctly calculated under script mode. Here is a minimal example to reproduce:

```python
# loop.py

import torch

class MyModule(torch.nn.Module):
    def __init__(self):
        super(MyModule, self).__init__()
        self.linear = torch.nn.Linear(4, 4, bias=False)
    def forward(self, xs):
        h = torch.zeros(3, 4)
        for i in range(xs.size(0)):
            h = torch.tanh(h + self.linear(xs[i]))
        return h

mymod = MyModule()
x = torch.rand(10, 3, 4)

y = mymod(x)
y.sum().backward()
print(mymod.linear.weight.grad)
mymod.linear.weight.grad.zero_()

traced_mod = torch.jit.trace(mymod, (x,))
y = traced_mod(x)
y.sum().backward()
print(traced_mod.linear.weight.grad)
traced_mod.linear.weight.grad.zero_()

script_mod = torch.jit.script(mymod, (x,))
y = script_mod(x)
y.sum().backward()
print(script_mod.linear.weight.grad)
print(script_mod.code)
script_mod.linear.weight.grad.zero_()
```

## To Reproduce

1. Run the above codes with `python loop.py`
1. Example output:
```
tensor([[12.2146, 10.0927, 12.0888, 10.9662],
        [ 1.9247,  1.9788,  2.3748,  2.6247],
        [ 0.9796,  0.6617,  0.8774,  0.7659],
        [ 9.4632,  7.9788, 10.1042,  9.0928]])
loop.py:9: TracerWarning: Converting a tensor to a Python index might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  for i in range(xs.size(0)):
tensor([[12.2146, 10.0927, 12.0888, 10.9662],
        [ 1.9247,  1.9788,  2.3748,  2.6247],
        [ 0.9796,  0.6617,  0.8774,  0.7659],
        [ 9.4632,  7.9788, 10.1042,  9.0928]])
/miniconda3/envs/ws/lib/python3.7/site-packages/torch/jit/__init__.py:1252: UserWarning: `optimize` is deprecated and has no effect. Use `with torch.jit.optimized_execution() instead
  warnings.warn("`optimize` is deprecated and has no effect. Use `with torch.jit.optimized_execution() instead")
tensor([[1.4874, 1.1861, 1.9228, 1.9684],
        [0.9283, 0.7348, 1.1052, 1.3079],
        [0.6291, 0.2742, 0.6101, 0.6308],
        [1.4516, 1.2129, 1.9343, 1.9622]])
def forward(self,
    xs: Tensor) -> Tensor:
  h = torch.zeros([3, 4], dtype=None, layout=None, device=None, pin_memory=None)
  h0 = h
  for i in range(torch.size(xs, 0)):
    _0 = (self.linear).forward(torch.select(xs, 0, i), )
    h0 = torch.tanh(torch.add(h0, _0, alpha=1))
  return h0
```

## Expected behavior

Script mode should give the same result as the native pytorch and trace mode.

## Environment

```
Collecting environment information...
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.0

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.13.3

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: Tesla V100-SXM2-16GB
Nvidia driver version: 418.87.01
cuDNN version: /usr/local/cuda-10.1/targets/x86_64-linux/lib/libcudnn.so.7.6.2

Versions of relevant libraries:
[pip3] numpy==1.15.4
[conda] blas                      1.0                         mkl
[conda] mkl                       2019.4                      243
[conda] mkl-service               2.3.0            py37he904b0f_0
[conda] mkl_fft                   1.0.15           py37ha843d7b_0
[conda] mkl_random                1.1.0            py37hd6b4f25_0
[conda] pytorch                   1.4.0           py3.7_cuda10.0.130_cudnn7.6.3_0    pytorch
```

## Additional context

It looks like it is because of the loop boundary being `xs.size[0]`. The code runs correctly if the loop boundary is a constant.

cc @suo

**Comment by suo** [comment]
Yes, this is a limitation of the tracing-based frontend to TorchScript. The warning in your repro explains why:
```
loop.py:9: TracerWarning: Converting a tensor to a Python index might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
```

An alternative is to use `torch.jit.script`, which will preserve control flow like loops.

Going to close this out to help with our issue tracking, but feel free to keeping commenting if you have more questions!

**Comment by jermainewang** [comment]
Hey @suo , I understand the issue with trace mode. However, the problem is with the **script** mode. Although it captures the for loop correctly, it does not compute the same gradient values as the other modes. Please double check.

**Comment by jermainewang** [comment]
@suo, any update on this?

**Comment by suo** [comment]
Sorry, I misread your issue! Re-opening, but I won't be able to look until tomorrow (today is a holiday for us)

**Comment by suo** [comment]
hmmmm, your problem does not reproduce for me on the latest master. Running your original repro, I get the same gradients for all three. Could you try installing the latest nightly and trying?

**Comment by rmcavoy** [comment]
@suo If you print the script.code which version of linear is it importing? As my tests above showed "The two script'd codes are exceedingly similar with the largest difference that I could spot being the non-fixed size (broken) loop imports torch.torch.nn.modules.linear and the fixed size (working) loop imports torch.torch.nn.modules.linear.___torch_mangle_2"

**Comment by suo** [comment]
This is what the original `script_mod.code` looks like: 
```
def forward(self,
    xs: Tensor) -> Tensor:
  h = torch.zeros([3, 4], dtype=None, layout=None, device=None, pin_memory=None)
  h0 = h
  for i in range(torch.size(xs, 0)):
    _0 = (self.linear).forward(torch.select(xs, 0, i), )
    h0 = torch.tanh(torch.add(h0, _0, alpha=1))
  return h0
```

It's very different from what yours looks like, which leads me to suspect you may be on an older version of PyTorch. What does `torch.__version__` say?

**Comment by chenzhekl** [comment]
Wang seems to be on `1.4.0` according to the environment he pasted.

**Comment by suo** [comment]
I still can't reproduce it, @jermainewang: could you try your repro on the nightly? Possibly it was fixed after the 1.4.0 branch was cut.

**Comment by jermainewang** [comment]
Tried on the latest build 1.5.0.dev20200203. It works like magic. Thanks for the help.
